src/frontend/src/optimizer/mod.rs

```diff
@@ -500,13 +500,42 @@ impl PlanRoot {
         Ok(plan)
     }
 
+    /// Auxiliary function to ensure there is **exchange directly before table scan**,
+    /// so that it will be pushed to compute node.
+    fn enforce_exchange_above_table_scan(plan: PlanRef) -> PlanRef {
+        if plan.node_type() != PlanNodeType::BatchExchange
+            && plan.node_type() != PlanNodeType::BatchSeqScan
+        {
+            let new_inputs = plan
+                .inputs()
+                .into_iter()
+                .map(|input| {
+                    if input.node_type() == PlanNodeType::BatchSeqScan
+                        && !has_batch_seq_scan_where(plan.clone(), |s| s.logical().is_sys_table())
+                    {
+                        let order = input.order().clone();
+                        BatchExchange::new(input, order, Distribution::Single).into()
+                    } else {
+                        Self::enforce_exchange_above_table_scan(input)
+                    }
+                })
+                .collect_vec();
+            plan.clone_with_inputs(&new_inputs)
+        } else {
+            plan
+        }
+    }
+
     /// Optimize and generate a batch query plan for local execution.
     pub fn gen_batch_local_plan(&mut self) -> Result<PlanRef> {
         let mut plan = self.gen_batch_plan()?;
 
         // Convert to local plan node
         plan = plan.to_local_with_order_required(&self.required_order)?;
 
+        // Ensure there is exchange before all seq scan.
+        plan = Self::enforce_exchange_above_table_scan(plan);
```

Comment:
Will this have some duplicated effects with `require_additional_exchange_on_root`? 🤔

Reply:
Yes`enforce_exchange_above_table_scan` should accomplish this too.

Will remove it.